### PR TITLE
expression: Fix the issue that cast str as datetime with time_zone results in loss of precision

### LINKED ISSUE #49555: Insert a special datetime with time_zone results in loss of precision
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t(a datetime(3));
insert into t values ('2020-01-01 12:00:00.123456 +0600 PST'); # The format is Golang’s default format
insert into t values ('2020-01-01 12:00:00.123456 -0600 PST');
select * from t;
```

### 2. What did you expect to see? (Required)

Got error during insert or keep result like `2020-01-01 12:00:00.123`

MySQL will report error during insert.

### 3. What did you see instead (Required)

```
mysql> select * from t;
+-------------------------+
| a                       |
+-------------------------+
| 2020-01-01 12:00:00.123 |
| 2020-01-01 12:00:00.000 |
+-------------------------+
2 rows in set (0.01 sec)
```

### 4. What is your TiDB version? (Required)

https://github.com/pingcap/tidb/commit/23953a7cbe82a18390f2a213777a57f48e203d18

### PR DESCRIPTION
**Pull request by SeaRise**
### What problem does this PR solve?

Issue Number: close #49555

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that cast str as datetime with time_zone results in loss of precision
```

### CODE REVIEW

File: pkg/types/time_test.go [hunk 1]
```diff
@@ -114,6 +114,8 @@ func TestDateTime(t *testing.T) {
 
 		// For issue 35291
 		{"2020-01-01 12:00:00.123456+05:00", "2020-01-01 07:00:00.123456"},
+		// For issue 49555
+		{"2020-01-01 12:00:00.123456-05:00", "2020-01-01 07:00:00.123456"},
```
Comment (Defined2014) [inline pkg/types/time_test.go:118]:
```suggestion
		{"2020-01-01 12:00:00.123456-05:00", "2020-01-01 17:00:00.123456"},
```

File: pkg/types/time.go [hunk 2]
```diff
@@ -587,7 +590,7 @@ func GetFracIndex(s string) (index int) {
 	}
 	index = -1
 	for i := end; i >= 0; i-- {
-		if unicode.IsPunct(rune(s[i])) {
+		if s[i] != '+' && s[i] != '-' && unicode.IsPunct(rune(s[i])) {
```
Comment (Defined2014) [inline pkg/types/time.go:593]: We have some functions in `types/time.go` and `types/helper.go`, like `isValidSeparator` and `isPunctuation`. Not sure could we directly use them.

Comment (SeaRise) [reply pkg/types/time.go]: It looks like there is no function available.
BTW, I replace `unicode.IsPunct` by `isPunctuation`.

Comment (Defined2014) [reply pkg/types/time.go]: Ok, just a question more than suggestion.

Seems `punctuation` in different language has its own meaning. I think `isPunctuation` function is followed `std::ispunct` which may different with `unicode.IsPunct`. Not sure which one is correct.

Comment (SeaRise) [reply pkg/types/time.go]: As `isPunctuation` was used consistently in time.go, I have replaced it here. It seems that in time.go, either all function should use `isPunctuation`, or all instances should use `unicode.IsPunct`.

Comment (SeaRise) [reply pkg/types/time.go]: I believe replacing it with `isPunctuation` is safe here, as this code snippet was introduced by https://github.com/pingcap/tidb/pull/9933 to handle the case like `2019.01.01 00:00:00`.

**Review by Defined2014 — APPROVED** [review]

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**